consider systime's difference between swarm and engine

allencloud opened the pull request:
fixed #2310 

This PR did:
1. add a field **DeltaDuration** in struct Engine;
2. use `swarm systime - docker systime in info api` to fill DeltaDuration
3. when calculate Status and Created time in `docker ps`, use DeltaDuration

Signed-off-by: allencloud <allen.sun@daocloud.io>

dongluochen commented on cluster/engine.go:
```diff
@@ -491,6 +492,14 @@ func (e *Engine) updateSpecs() error {
+
+	if info.SystemTime != "" {
+		engineTime, _ := time.Parse(time.RFC3339Nano, info.SystemTime)
+		e.DeltaDuration = time.Now().UTC().Sub(engineTime)
```
This delta is an estimation of time difference between manager and engine with adjustment of delays (Engine response delay + network delay + manager process delay). The adjustment of delays is not a constant. If the servers are sync up on time, this delta might be the source of error. 

Another way to fix is to output time difference as error in log and in `docker ps`, but without adjusting time. 

I think both approaches are ok.

dongluochen replied:
I'm thinking if it's better to take a timestamp after line 452, only if time difference is larger than a threshold (1 or 2 seconds), use the DeltaDuration.

allencloud replied:
I will add the threshold in updateSpecs. So Swarm will not enter an if when receiving a request

dongluochen commented on cluster/container.go:
```diff
@@ -44,17 +44,21 @@ func StateString(state *types.ContainerState) string {
+			return fmt.Sprintf("Restarting (%d) %s ago", state.ExitCode, units.HumanDuration(time.Now().UTC().Sub(finishedAt.Add(duration))))
 		}
-		return fmt.Sprintf("Up %s", units.HumanDuration(time.Now().UTC().Sub(startedAt)))
+		return fmt.Sprintf("Up %s", units.HumanDuration(time.Now().UTC().Sub(startedAt.Add(duration))))
```
A check is needed here to make sure "time.Now().UTC().Sub(startedAt.Add(duration)))" is not in the future. A container cannot be `up -60 seconds`, i.e., in the future.

dongluochen replied:
I suspect this wouldn't happen :) But please test.

allencloud replied:
Yes, it is reasonable. Today I will polish this PR.

allencloud commented:
This PR needs to change engine_test.go, while I found test in this file is always in a single line and is not so readable. As a result I changed a lot in format of engine_test.go. @nishanttotla 

I added this delta in the raw data Info in updateContainer, so I do not have to change codes in container.go and handler.go. @dongluochen

allencloud commented:
ping @dongluochen @nishanttotla @vieux @jimmyxian

dongluochen commented:
@allencloud Please rebase.

allencloud commented:
@dongluochen Rebased. PTAL

dongluochen commented:
LGTM

abronan commented:
LGTM
